test/jdk/java/util/zip/ZipFile/InvalidBytesInEntryNameOrComment.java

```diff
@@ -21,163 +21,304 @@
      */
-    private byte[] templateZIP() throws IOException {
-        ByteArrayOutputStream bout = new ByteArrayOutputStream();
-        try (ZipOutputStream zo = new ZipOutputStream(bout)) {
-            ZipEntry commentEntry = new ZipEntry("file");
-            commentEntry.setComment("Comment");
-            zo.putNextEntry(commentEntry);
+    @Test
+    public void traverseZipWithZipInputStreamTest() throws IOException {
+        Files.write(ZIP_FILE, zipArray);
+        try( ZipInputStream zis = new ZipInputStream(new FileInputStream(ZIP_FILE.toFile()))) {
+            ZipEntry ze;
+            while((ze = zis.getNextEntry()) != null ) {
+                System.out.printf("Entry: %s%n", ze.getName());
+            }
         }
-        return bout.toByteArray();
     }
 
     /**
-     * Make a ZIP with invalid bytes in the CEN name field
+     * Validate that a ZipException is thrown when an entry name or entry comment
+     * within a CEN file header contains an invalid UTF-8 byte sequence.
+     *
+     * @param offset the offset to the file name or file comment within the CEN
+     *               file header
+     * @throws IOException if an error occurs
      */
-    private Path invalidName(String name, byte[] template) throws IOException {
-        ByteBuffer buffer = copyTemplate(template);
-        int off = cenStart(buffer);
-        // Name field starts here
-        int noff = off + CEN_HDR;
-
-        // Write invalid bytes
-        buffer.put(noff, INVALID_UTF8_BYTE_SEQUENCE, 0, INVALID_UTF8_BYTE_SEQUENCE.length);
-        return writeFile(name, buffer);
-
+    @ParameterizedTest
+    @MethodSource("CENCommentOffsets")
+    public void testInValidEntryNameOrComment(int offset)
+            throws IOException {
+        createInvalidUTFEntryInZipFile(offset);
+        Throwable ex = assertThrows(ZipException.class, () ->
+                {
+                    try (ZipFile zf = new ZipFile(ZIP_FILE.toFile())) {};
+                }
+        );
+        assertEquals(CEN_BAD_ENTRY_NAME_OR_COMMENT, ex.getMessage());
     }
-
     /**
-     * Make a copy of the ZIP template and wrap it in a little-endian
-     * ByteBuffer
+     * Validate that a null is returned from ZipFile::getComment when the
+     * comment contains an invalid UTF-8 byte sequence.
+     * @throws IOException if an error occurs
      */
-    private ByteBuffer copyTemplate(byte[] template) {
-        return ByteBuffer.wrap(Arrays.copyOf(template, template.length))
-                .order(ByteOrder.LITTLE_ENDIAN);
+    @Test
+    public void testInValidZipFileComment() throws IOException {
+        createInvalidUTFEntryInZipFile(ZIP_FILE_COMMENT_OFFSET);
+        try (ZipFile zf = new ZipFile(ZIP_FILE.toFile())) {
+            assertNull(zf.getComment());
+        }
     }
 
     /**
-     * Make a ZIP with invalid bytes in the CEN comment field
+     * Validate that a ZipException is thrown when an entry name is encountered
```

Comment:
Is the "is" here out of place or are we missing a "which"?

```suggestion
     * Validate that a ZipException is thrown when an entry name encountered
```